lib/browser/source.js

```diff
@@ -11,7 +11,95 @@ class FileSource {
   close() {}
 }
 
-export function getSource(input) {
+class StreamSource {
+  constructor(reader, chunkSize) {
+    // Ensure that chunkSize is an integer and not something else or Infinity.
+    chunkSize = +chunkSize;
+    if (!isFinite(chunkSize)) {
+      throw new Error("cannot create source for stream without a finite value for the `chunkSize` option");
+    }
+    this._chunkSize = chunkSize;
+
+    this._buffer = undefined;
+    this._bufferOffset = 0;
+    this._reader = reader;
+    this._done = false;
+  }
+
+  slice(start, end, callback) {
+    if (start < this._bufferOffset) {
+      throw new Error("requested data is before the reader's current offset");
+    }
+
+    return this._readUntilEnoughDataOrDone(start, end, callback);
+  }
+
+  _readUntilEnoughDataOrDone(start, end, callback) {
+    const hasEnoughData = end <= this._bufferOffset + len(this._buffer);
+    if (this._done || hasEnoughData) {
+      callback(this._getDataFromBuffer(start, end));
+      return;
+    }
+    this._reader.read().then(({ value, done }) => {
+      if (done) {
+        this._done = true;
+      } else if (this._buffer === undefined){
+        this._buffer = value;
+      } else {
+        this._buffer = concat(this._buffer, value);
+      }
+
+      this._readUntilEnoughDataOrDone(start, end, callback);
+    });
+  }
+
+  _getDataFromBuffer(start, end) {
+    const hasAllDataBeenRead = start >= this._bufferOffset + len(this._buffer);
+    if (this._done && hasAllDataBeenRead) {
+      return null;
+    }
+    const bufferToReturn = this._buffer.slice(
+      start - this._bufferOffset,
+      end - this._bufferOffset
+    );
+    const chunkSize = end - start;
+    if (len(this._buffer) >= 2 * chunkSize) {
+      this._buffer = this._buffer.slice(chunkSize);
+      this._bufferOffset += chunkSize;
+    }
+    return bufferToReturn;
+  }
+
+  close() {}
+}
+
+function len(blobOrArray) {
+  if (blobOrArray === undefined) return 0;
+  if (blobOrArray.size !== undefined) return blobOrArray.size;
+  return blobOrArray.length;
+}
+
+/*
+  Typed arrays and blobs don't have a concat method.
+  This function helps StreamSource accumulate data to reach chunkSize.
+*/
+function concat(a, b) {
+  if (a.concat) {
+    return a.concat(b);
+  }
+  if (a instanceof Blob) {
+    return new Blob([a,b], {type: a.type});
+  }
+  if (a.set) {
```

Comment:
For which data type is this condition? Could you add a comment?

Reply:
This is for typed arrays. I added a comment.